source/common/stats/stat_merger.cc

```diff
@@ -0,0 +1,114 @@
+      // Once the child is up and reporting stats, its own control plane state and liveness is what
+      // we're interested in.
+      {std::regex(".*\\.control_plane.connected_state$"), Gauge::CombineLogic::NoImport},
+      {std::regex("^server.live$"), Gauge::CombineLogic::NoImport},
+      // Properties that should reasonably have some continuity across hot restart. The parent's
+      // last value should be a relatively accurate starting point, and then the child can update
+      // from there when appropriate. (All of these exceptional stats used with set() rather than
+      // add()/sub(), so the child's new value will in fact overwrite.)
+      {std::regex("^cluster_manager.active_clusters$"),
+       Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex("^cluster_manager.warming_clusters$"),
+       Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex("^cluster\\..*\\.membership_total$"),
+       Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex("^cluster\\..*\\.membership_healthy$"),
+       Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex("^cluster\\..*\\.membership_degraded$"),
+       Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex("^cluster\\..*\\.max_host_weight$"),
+       Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex(".*\\.total_principals$"), Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex("^listener_manager.total_listeners_active$"),
+       Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      {std::regex("^overload\\..*\\.pressure$"), Gauge::CombineLogic::OnlyImportWhenUnusedInChild},
+      // Due to the fd passing, the parent's view of whether its listeners are in transitive states
+      // is not useful.
+      {std::regex("^listener_manager.total_listeners_draining$"), Gauge::CombineLogic::NoImport},
+      {std::regex("^listener_manager.total_listeners_warming$"), Gauge::CombineLogic::NoImport},
+      // Static properties known at startup.
+      {std::regex("^server.concurrency$"), Gauge::CombineLogic::NoImport},
+      {std::regex("^server.hot_restart_epoch$"), Gauge::CombineLogic::NoImport},
+      {std::regex("^runtime.admin_overrides_active$"), Gauge::CombineLogic::NoImport},
+      {std::regex("^runtime.num_keys$"), Gauge::CombineLogic::NoImport},
+  };
+  for (const auto& exception : *nonstandard_combine_logic) {
+    std::smatch match;
+    if (std::regex_match(gauge_name, match, exception.first)) {
+      gauge.setCombineLogic(exception.second);
+      return exception.second;
+    }
+  }
+  gauge.setCombineLogic(Gauge::CombineLogic::Accumulate);
+  return Gauge::CombineLogic::Accumulate;
+}
+
+void StatMerger::mergeCounters(const Protobuf::Map<std::string, uint64_t>& counter_deltas) {
+  for (const auto& counter : counter_deltas) {
+    target_store_.counter(counter.first).add(counter.second);
```

Comment:
I think we have another interesting potential issue here. I think it's now possible for stats to "leak" between process restart. This can happen if a counter from an old process gets written that never exists in the new process. The counter will be created here, but not actually owned by any scope, and thus if it is never created in the new process, it could effectively sit around forever even though it is never referenced by anything.

In the old code, so that we are on the same page, this was not an issue, because we would write into shared memory, the new process would never "attach" thus abandoning the stat. Then the old process would shut down and the stat would be freed from shared memory.

I will need to think about this more as I'm not sure what the best solution is in this new world. The simplest solution would be to just not add to counters that don't exist yet in the new process. However, if we do this, and we are sending deltas, we may actually lose data. The other option would be to somehow store deltas for unknown counters through the hot restart drain period, and then if they are still not attached by the time the parent shuts down, we just abandon the counter deltas.

I'm not sure we need to solve this in this PR (meaning I think we can ship with this issue) but I think we should have a giant TODO and tracking issue to cover this case. @fredlas @jmarantz would love to get both of your opinions on this one.

Reply:
I like your last idea: drop the unattached stats at the end of hot-restart drain, and agree it could be a follow-up.

Reply:
Quick note for TODO/issue follow up, this will also apply to gauges (in terms of leaking) so please comment/track there also.

Reply:
TODO added. Maybe a clean way would be to let everything be the way it currently is in this PR, but add a flag bit "used by this process" that would be set by any interaction (other than merging) with the stat. Then, at the end of the hot restart, get rid of all stats without that flag set.

Reply:
The only thing I don't like about the flag approach is there would be a performance hit on every access to set this flag I think. This will need some thinking, but I think we can come up with a solution that is some intersection of the suggestions already proposed.

Reply:
That makes sense, although I wonder if it's possible to repurpose the existing Used flag, which is already getting set by all stat modifications, and which currently (going by my not-too-deep look just now) doesn't actually appear to be doing all that much. I've also been thinking about how to implement the "store unknown deltas" idea, though; I'm sure that will be doable. Anyways, I will copy all of this over to a new issue.